packages/candid/src/idl.ts

```diff
@@ -2032,3 +2069,174 @@ export function Func(args: Type[], ret: Type[], annotations: string[] = []): Fun
 export function Service(t: Record<string, FuncClass>): ServiceClass {
   return new ServiceClass(t);
 }
+
+/**
+ * The list of relations between types we assume to hold. Uses the types .name property as key
+ */
+class Relations {
+  private rels: Map<string, Map<string, boolean>>;
+
+  constructor(relations = new Map()) {
+    this.rels = relations;
+  }
+
+  copy(): Relations {
+    const copy = new Map();
+    for (const [key, value] of this.rels.entries()) {
+      const valCopy = new Map(value);
+      copy.set(key, valCopy);
+    }
+    return new Relations(copy);
+  }
+
+  /// Returns whether we know for sure that a relation holds or doesn't (`true` or `false`), or
+  /// if we don't know yet (`undefined`)
+  known(t1: Type, t2: Type): boolean | undefined {
+    return this.rels.get(t1.name)?.get(t2.name);
+  }
+
+  addNegative(t1: Type, t2: Type) {
+    this.addNames(t1.name, t2.name, false);
+  }
+
+  add(t1: Type, t2: Type) {
+    this.addNames(t1.name, t2.name, true);
+  }
```

Comment:
Maybe we shouldn't worry but t.name is a property that is recomputed on each access.  It's quite expensive for record and variants. Do we want to address this, or leave it as an issue? It's a shame we can't use something else to key on.